### PR TITLE
Fix body reader of http backend

### PR DESCRIPTION
**Pull request by sundowndev**
| Q                 | A
| ----------------- | ---
| 🐛 Bug fix?       | no
| 🚀 New feature?   | no
| ⚠ Deprecations?   | no
| ❌ BC Break       | no
| 🔗 Related issues | #...
| ❓ Documentation  | no 

## Description

As discussed in #406 , we should mock the http.Client instead of running real HTTP calls on the network.

This change also fixes an issue we recently discovered: the response body is always empty because a buffer was reading the body content already.

### CODE REVIEW

**Review by wbeuil — DISMISSED** [review]
LGTM

**Review by wbeuil — DISMISSED** [review]
LGTM, @eliecharra @moadibfr ping on this one.

File: pkg/iac/terraform/state/backend/http_reader.go [hunk 1]
```diff
@@ -17,7 +17,11 @@ type HTTPBackend struct {
 	reader io.ReadCloser
 }
 
-func NewHTTPReader(rawURL string, opts *Options) (*HTTPBackend, error) {
+type HttpClient interface {
```
Comment (eliecharra) [inline pkg/iac/terraform/state/backend/http_reader.go:20]: This should be put in `test/http.go` with the mock file alongside (mockery default config)

File: pkg/iac/terraform/state/backend/http_reader.go [hunk 2]
```diff
@@ -27,25 +31,23 @@ func NewHTTPReader(rawURL string, opts *Options) (*HTTPBackend, error) {
...
 
-	client := &http.Client{}
 	res, err := client.Do(req)
 	if err != nil {
 		return nil, err
 	}
 
-	buf := new(bytes.Buffer)
-	_, err = buf.ReadFrom(res.Body)
+	body, err := io.ReadAll(res.Body)
```
Comment (eliecharra) [inline pkg/iac/terraform/state/backend/http_reader.go:39]: `client.Do` and response read should be done in the `Read` func IMHO, it's kinda unexpected to have this kind of logic within a `New...` method.

File: pkg/iac/terraform/state/backend/http_reader_test.go [hunk 3]
```diff
@@ -15,10 +19,11 @@ func TestNewHTTPReader(t *testing.T) {
...
 	}
 	tests := []struct {
-		name    string
-		args    args
-		wantURL string
-		wantErr error
+		name       string
+		args       args
+		wantErr    error
+		httpClient func() HttpClient
```
Comment (eliecharra) [inline pkg/iac/terraform/state/backend/http_reader_test.go:25]: You can put `HttpClient` directly and in each case you can do `httpClient: func() HttpClient { return &http.Client{} }()`

This more homogenous with what we do in other tests

**Review by eliecharra — CHANGES_REQUESTED** [review]

File: pkg/iac/terraform/state/backend/http_reader_test.go (hunk 3 above)
Comment (sundowndev) [reply pkg/iac/terraform/state/backend/http_reader_test.go:25]: Isn't it what's already done in this PR ?

File: pkg/iac/terraform/state/backend/http_reader.go [hunk 4]
```diff
@@ -27,30 +31,23 @@ func NewHTTPReader(rawURL string, opts *Options) (*HTTPBackend, error) {
...
 }
 
 func (h *HTTPBackend) Read(p []byte) (n int, err error) {
 	if h.reader == nil {
-		return 0, errors.New("Reader not initialized")
+		res, err := h.client.Do(h.request)
+		if err != nil {
+			return 0, err
+		}
+		h.reader = res.Body
```
Comment (eliecharra) [inline pkg/iac/terraform/state/backend/http_reader.go:43]: I don't think we need the reader to be a struct attribute anymore as we do all the job in the `Read` now

File: pkg/iac/terraform/state/backend/http_reader_test.go [hunk 5]
```diff
@@ -28,52 +32,91 @@ func TestNewHTTPReader(t *testing.T) {
...
+
+				return m
+			}(),
+			expected: "{}",
 		},
 		{
 			name: "Should fail with bad status code",
 			args: args{
-				url: "https://raw.githubusercontent.com/cloudskiff/driftctl-test/main/.dockerignore",
+				url: "https://wrong.url/cloudskiff/driftctl/main/terraform.tfstate",
```
Comment (eliecharra) [inline pkg/iac/terraform/state/backend/http_reader_test.go:74]: This is very minor, but you can use `example.com` for this purpose, it is what it has been [designed for](https://www.iana.org/domains/reserved).

File: pkg/iac/terraform/state/backend/http_reader_test.go [hunk 6]
```diff
@@ -107,82 +150,12 @@ func TestHTTPBackend_Close(t *testing.T) {
 	for _, tt := range tests {
 		t.Run(tt.name, func(t *testing.T) {
 			h := &HTTPBackend{
-				url:    tt.fields.url,
-				reader: tt.fields.reader(),
+				request: tt.fields.req,
+				reader:  tt.fields.reader(),
```
Comment (eliecharra) [inline pkg/iac/terraform/state/backend/http_reader_test.go:154]: Should be `reader:  tt.fields.reader` no ? (without func call `()`)

**Review by eliecharra — CHANGES_REQUESTED** [review]

File: pkg/iac/terraform/state/backend/http_reader.go (hunk 4 above)
Comment (sundowndev) [reply pkg/iac/terraform/state/backend/http_reader.go:43]: What should be the purpose of Close() method then?

File: pkg/iac/terraform/state/backend/http_reader_test.go (hunk 6 above)
Comment (eliecharra) [reply pkg/iac/terraform/state/backend/http_reader_test.go:154]: > But I can turn this field into a io.ReadCloser type instead

Yeah it is the same remark as for the `HttpClient`

File: pkg/iac/terraform/state/backend/http_reader.go (hunk 1 above)
Comment (sundowndev) [reply pkg/iac/terraform/state/backend/http_reader.go:20]: That'd make the HttpClient become an interface for testing, but it's actually used in the backend package. I don't think we want to use a test interface in our code : `func NewHTTPReader(client test.HttpClient...`, do we?

File: pkg/iac/terraform/state/backend/http_reader.go (hunk 4 above)
Comment (eliecharra) [reply pkg/iac/terraform/state/backend/http_reader.go:43]: You are right I forgot the `ReadCloser` interface nevermind

File: pkg/iac/terraform/state/backend/http_reader.go (hunk 1 above)
Comment (eliecharra) [reply pkg/iac/terraform/state/backend/http_reader.go:20]: That's true, the fact is that this generic interface should not be in `pkg/iac/terraform/state/backend/`.

Maybe we can move it to a new `pkg/http` package ?

File: mocks/http.go [hunk 7]
```diff
@@ -0,0 +1,19 @@
+package mocks
+
+import (
+	"net/http"
+
+	mock "github.com/stretchr/testify/mock"
+)
+
+// HttpClient is an autogenerated mock type for the Client type
+type HttpClient struct {
```
Comment (eliecharra) [inline mocks/http.go:10]: The mock should be in `pkg/http` according to our mockery configuration.

**Review by eliecharra — CHANGES_REQUESTED** [review]

**Review by eliecharra — APPROVED** [review]

**Pull request merged**